### PR TITLE
[Feature:Submission] Display randomized notebooks

### PR DESCRIPTION
**Pull request by shailpatels**
### Please check if the PR fulfills these requirements:

* [ ] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Notebooks can only be created a course specific level.

### What is the new behavior?
This PR begins work on creating user specific exams (notebooks) by being able to parse item pools. 

A new model is created to hold some of the data part of it is still tangled up in some of the other views models.

Autograding does not work with this PR currently and will take a bit more work than I initially thought so that'll be done in a seperate PR. 

`submit.notebook` file is created under the student submissions dir containing the hashes and item pools selected for their notebook 

### Other information?

Should be backward compatible with existing configs.

The config generated by main_configure does not include the item pool, php is parsing part of full_config to get the item pool this should be ok for now but currently item pool is not being validated

If a notebook has the potential to select the same item pool to pick from multiple times a warning banner is displayed to instructors only.

image

### CODE REVIEW

**Comment by shailpatels** [comment]
@holzbh On Wednesday's meeting it was decided that instead of figuring out a method to avoid picking multiple questions we would instead show a warning to the instructor (maybe also to any grader like you mentioned).  This was done so the PR could get made & reviewed quicker since we would still have to think of how to handle the case when we do pick duplicate questions. 

As for the changing config , making sure the `initial_value` field is defined should be a quick fix, I'm not sure if it'll make sense to notify the student if a question has changed. I think a better solution would be to notify the student something like:
> You did not answer x questions, are you sure you want to submit?

Whenever a question is left blank
This what LMS does the last time I took an exam on that. 

Once this PR is merged and tested a bit more we can work on handling duplicate questions/autograding any other last-minute features - plenty of time for a hackathon before finals right :wink:

**Comment by holzbh** [comment]
I think I'm okay approving this. Last thing I noticed - even if `item_pool_selected` and `hashes` are empty (like in a traditional notebook gradeable like `notebook_basic`), we still write the .submit.notebook file. It's pretty small - 51 bytes per submission, and it's probably better to just have the file for any notebook (especially if a notebook goes from non-item-pool to item-pool I don't want things to break - and I could see an instructor doing this to preview all questions) so that we don't have surprise crashes. If someone thinks the 51 bytes/submission is going to add up to be problematic, we can discuss it.

File: site/app/models/notebook/UserSpecificNotebook.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+        $this->gradeable_id = $gradeable_id;
+        $this->user_id = $user_id;
+
+        $this->notebook_config = $this->replaceNotebookItemsWithQuestions($details);
+    }
+
+
+    //collect items from a notebook and replace them with the actual
+    //notebook values
```
Comment (MasterOdin) [inline site/app/models/notebook/UserSpecificNotebook.php:63]:
```suggestion
/**
 * Collect the items from a notebook and replace them with the actual values
 */
```

File: site/app/models/notebook/UserSpecificNotebook.php [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    private function replaceNotebookItemsWithQuestions($raw_notebook) {
```
Comment (MasterOdin) [inline site/app/models/notebook/UserSpecificNotebook.php:64]:
```suggestion
    private function replaceNotebookItemsWithQuestions(array $raw_notebook): array {
```

File: site/app/models/notebook/UserSpecificNotebook.php [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        $hash = hexdec(substr(md5("{$item_label}|{$gid}|{$uid}|{$semester}|{$course}"), 24, 8));
+
+        $selected = $hash % $from_pool_count;
+        $this->hashes[] = $hash;
+
+        return $selected;
+    }
+
+
+    private function searchForItemPool($tgt_name) {
```
Comment (MasterOdin) [inline site/app/models/notebook/UserSpecificNotebook.php:130]:
```suggestion
    private function searchForItemPool(string $tgt_name): array {
```

File: site/app/models/notebook/UserSpecificNotebook.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    private function getItemFromPool($item) {
+        $item_label = $item['item_label'];
+        $selected = $this->getNotebookHash($item_label, count($item['from_pool']));
+        $item_from_pool = $item['from_pool'][$selected];
+        $this->selected_questions[] = $item_from_pool;
+
+        return $item_from_pool;
+    }
+
+    private function getNotebookHash($item_label, $from_pool_count) {
```
Comment (MasterOdin) [inline site/app/models/notebook/UserSpecificNotebook.php:113]:
```suggestion
    private function getNotebookHash(string $item_label, int $from_pool_count): int {
```

File: site/app/models/notebook/UserSpecificNotebook.php [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        if (count($multiple_selected) > 0) {
+            $seen = implode($multiple_selected, ",");
+            $this->warning = "Warning: The item_pool \"{$seen}\" has been selected multiple times, this could lead to notebooks with the same question given more than once.";
+        }
+
+        return $new_notebook;
+    }
+
+
+    private function getItemFromPool($item) {
```
Comment (MasterOdin) [inline site/app/models/notebook/UserSpecificNotebook.php:104]:
```suggestion
    private function getItemFromPool(array $item): string {
```

File: site/app/models/notebook/UserSpecificNotebook.php [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    protected $hashes = [];
+    /** @prop @var array of item_pool names selected */
+    protected $selected_questions = [];
+    /** @prop @var string warning if this notebook has potentially overlapping questions picked */
+    protected $warning = null;
+
+    private $gradeable_id;
+    private $user_id;
+
+    public function __construct(Core $core, array $details, $gradeable_id, $user_id) {
```
Comment (MasterOdin) [inline site/app/models/notebook/UserSpecificNotebook.php:40]:
```suggestion
    public function __construct(Core $core, array $details, ?string $gradeable_id, ?string $user_id) {
```

File: site/app/models/gradeable/AutogradingConfig.php [hunk 7]
```diff
@@ -283,8 +363,31 @@ public function getInputs() {
         return $this->inputs;
     }
 
-    public function getNotebook() {
-        return $this->notebook;
+    public function getNotebook($gradeable_id = null, $user_id = null) {
```
Comment (MasterOdin) [inline site/app/models/gradeable/AutogradingConfig.php:366]:
```suggestion
    public function getNotebook(?string $gradeable_id = null, ?string $user_id = null): UserSpecificNotebook {
```

File: site/app/models/gradeable/AutogradingConfig.php [hunk 8]
```diff
@@ -145,65 +154,98 @@ public function __construct(Core $core, array $details) {
...
+            }
+            else {
+                $this->part_names[$i] = "Part " . $i;
+            }
+        }
+    }
+
+
+    //TODO: remove once we can store notebook in its own model
+    public function parseNotebook($details) {
```
Comment (MasterOdin) [inline site/app/models/gradeable/AutogradingConfig.php:188]:
```suggestion
    public function parseNotebook(array $details): void {
```

File: site/app/models/gradeable/GradedGradeable.php [hunk 9]
```diff
@@ -276,17 +277,13 @@ public function getOverriddenComment() {
      * then 'recent_submission' is populated with 'initial_value' if one exists, otherwise it will be
      * blank.
      */
-    public function getUpdatedNotebook() {
-
-        // Get notebook
-        $newNotebook = $this->getGradeable()->getAutogradingConfig()->getNotebook();
-
+    public function getUpdatedNotebook($newNotebook) {
```
Comment (MasterOdin) [inline site/app/models/gradeable/GradedGradeable.php:280]:
```suggestion
    public function getUpdatedNotebook(array $newNotebook): array {
```

File: site/app/controllers/student/SubmissionController.php [hunk 10]
```diff
@@ -917,6 +918,20 @@ public function ajaxUploadSubmission($gradeable_id, $merge = null, $clobber = nu
...
+            $notebook = $gradeable->getAutogradingConfig()->getNotebook($gradeable_id, $who_id);
+
+            //save the notebook hashes and item selected
+            $json = [
+                "hashes" => $notebook->getHashes(),
+                "item_pools_selected" => $notebook->getSelectedQuestions()
+            ];
+
+            $tgt = FileUtils::joinPaths($version_path, ".submit.notebook");
+            file_put_contents($tgt, FileUtils::encodeJson($json));
```
Comment (MasterOdin) [inline site/app/controllers/student/SubmissionController.php:932]:
```suggestion
            FileUtils::writeJsonFile(FileUtils::joinPaths($version_path, ".submit_notebook"), $json);
```

File: site/app/models/gradeable/AutogradingConfig.php [hunk 11]
```diff
@@ -56,7 +59,7 @@ class AutogradingConfig extends AbstractModel {
     /** @prop @var AbstractGradingInput[] Grading input configs for all new types of gradeable input*/
     private $inputs = [];
     /** @prop @var AutogradingTestcase[] Cut-down information about autograding test cases*/
-    private $testcases = [];
+    public $testcases = [];
```
[the remark this answers is not in the document]
Comment (shailpatels) [reply site/app/models/gradeable/AutogradingConfig.php:62]: oh good catch, I'll revert that back - made public when testing

File: site/app/templates/generic/Banner.twig [hunk 12]
```diff
@@ -0,0 +1,13 @@
+{% import _self as self %}
+
+{# WRAP THE LATE DAY INFORMATION IN A DIV #}
```
[the remark this answers is not in the document]
Comment (shailpatels) [reply site/app/templates/generic/Banner.twig:3]: copied that from the twig late-day banner, will remove

**Comment by shailpatels** [comment]
@MasterOdin  thank you for the code review I'll implement the requested changes and suggestions, a quick question on PHP type hints, from now should we try and always use type hints on the parameters and return values, this a PHP feature I'm unfamiliar with.  Is this also something that can be linted?

@holzbh I'll make the quick fix for the `initial_value` warning getting thrown and yes I agree always leaving the file for a notebook gradeable will make things simpler (on the autograding side too as well). I imagine we'll use more of this file in the future when we add new features as well.

**Comment by MasterOdin** [comment]
> @MasterOdin thank you for the code review I'll implement the requested changes and suggestions, a quick question on PHP type hints, from now should we try and always use type hints on the parameters and return values, this a PHP feature I'm unfamiliar with. Is this also something that can be linted?

It is something I am going to increasingly push, yes. It improves the usefulness of phpstan and allows us to be more confident about some of our changes will not have unintended side effects. Once the codebase is increasingly typed, I do plan to add lint rules around this, but until we're closer to that objective, it's not possible.

**Comment by bmcutler** [comment]
Short answer boxes in the original notebook config style are broken.  The contents are wiped and the submitted files are empty.

File: site/app/models/notebook/UserSpecificNotebook.php (hunk 3 above)
Comment (bmcutler) [reply site/app/models/notebook/UserSpecificNotebook.php:130]: @shailpatels @MasterOdin 
Status of this requested change?

File: site/app/models/notebook/UserSpecificNotebook.php (hunk 4 above)
Comment (bmcutler) [reply site/app/models/notebook/UserSpecificNotebook.php]: @shailpatels @MasterOdin 
Status of this requested change?

File: site/app/controllers/student/SubmissionController.php [hunk 13]
```diff
@@ -1011,20 +1026,21 @@ public function ajaxUploadSubmission($gradeable_id, $merge = null, $clobber = nu
...
                 }
 
-                // FIXME: add error checking
+                //FIXME: add error checking
                 $file = fopen($dst, "w");
-
                 foreach ($answers as $answer_val) {
                     fwrite($file, $answer_val . "\n");
                 }
                 fclose($file);
```
Comment (MasterOdin) [inline site/app/controllers/student/SubmissionController.php:1034]:
```suggestion
                file_put_contents($dst, implode("\n", $answers) . "\n");
```

**Pull request merged**